### PR TITLE
ML-2648: Redis cluster support

### PR DESCRIPTION
**Pull request by alxtkr77**

### CODE REVIEW

**Review by gtopper — CHANGES_REQUESTED** [review]
Please revert the changes to code style. We can consider such changes, but not as part of a PR to add redis cluster support.

File: storey/redis_driver.py [hunk 1]
```diff
@@ -104,13 +107,9 @@ def redis(self):
...
         return self._redis
 
-    async def close(self):
-        if self._redis:
-            self._redis.close()
-
     @staticmethod
     def make_key(key_prefix, *parts):
-        return f"{key_prefix}{''.join([str(p) for p in parts])}"
+        return f"{{{key_prefix}{''.join([str(p) for p in parts])}}}"
```
Comment (gtopper) [inline storey/redis_driver.py:104]: What happened here? This change doesn't look right.

File: storey/redis_driver.py [hunk 2]
```diff
@@ -70,10 +70,13 @@ def __init__(self,
...
         if redis_client is not None:
             self._redis = redis_client
+            if isinstance(redis_client, redis.cluster.RedisCluster):
+                redis_type = RedisType.CLUSTER
+            if isinstance(redis_client, redis.Redis):
+                redis_type = RedisType.STANDALONE
         else:
             NeedsRedisAccess.__init__(self, redis_url)
             self._redis = None
-
```
Comment (gtopper) [inline storey/redis_driver.py:78]: I think this newline is actually appropriate here.

File: storey/redis_driver.py [hunk 3]
```diff
@@ -70,10 +70,13 @@ def __init__(self,
         # upon demand (any access to self.redis)
         if redis_client is not None:
             self._redis = redis_client
+            if isinstance(redis_client, redis.cluster.RedisCluster):
+                redis_type = RedisType.CLUSTER
+            if isinstance(redis_client, redis.Redis):
```
Comment (gtopper) [inline storey/redis_driver.py]:
```suggestion
            elif isinstance(redis_client, redis.Redis):
```

File: integration/test_redis_specific.py [hunk 4]
```diff
@@ -45,13 +54,22 @@ def test_redis_driver_write(redis):
...
+
     driver = RedisDriver(redis)
-    table = Table('test', driver)
+    table = Table(table_name, driver)
+    table_name = f'{table_name}/'
 
     # Create the data we'll join with in Redis.
-    driver.redis.hset("storey:test/2:static", mapping={"name": "1234"})
+    hash_key = RedisDriver.make_key('storey:', table_name, '2')
+    redis_key= RedisDriver._static_data_key(hash_key)
```
Comment (gtopper) [inline integration/test_redis_specific.py]:
```suggestion
    redis_key = RedisDriver._static_data_key(hash_key)
```

File: integration/test_redis_specific.py [hunk 5]
```diff
@@ -68,3 +86,6 @@ def test_redis_driver_join(redis):
         {'col1': 1, 'col2': '2', 'name': 1234}]
 
     assert termination_result == expected_result
+
+    for key in driver.redis.scan_iter(f'*storey:{table_name}*'):
+        driver.redis.delete(key)
```
Comment (gtopper) [inline integration/test_redis_specific.py]: If the point is cleanup, better put this in teardown or in a finally clause.

File: integration/test_redis_specific.py [hunk 6]
```diff
@@ -45,13 +54,22 @@ def test_redis_driver_write(redis):
 
     assert data_strings == {"col1": '0'}
 
+    for key in driver.redis.scan_iter(f'*storey:{table_name}*'):
+        driver.redis.delete(key)
 
 def test_redis_driver_join(redis):
+    random_name = ''.join(random.choice(string.ascii_lowercase) for i in range(5))
```
Comment (gtopper) [inline integration/test_redis_specific.py]: Avoid using random without seed in tests as it makes tests nondeterministic.

File: integration/test_redis_specific.py [hunk 7]
```diff
@@ -45,13 +54,22 @@ def test_redis_driver_write(redis):
 
     assert data_strings == {"col1": '0'}
 
+    for key in driver.redis.scan_iter(f'*storey:{table_name}*'):
+        driver.redis.delete(key)
```
Comment (gtopper) [inline integration/test_redis_specific.py]: If the point is cleanup, better put this in teardown or in a finally clause.

File: integration/test_redis_specific.py [hunk 8]
```diff
@@ -25,17 +27,24 @@ def redis():
     return get_redis_client()
 
 def test_redis_driver_write(redis):
+    random_name = ''.join(random.choice(string.ascii_lowercase) for i in range(5))
```
Comment (gtopper) [inline integration/test_redis_specific.py]: Avoid using random without seed in tests as it makes tests nondeterministic.

File: storey/redis_driver.py (hunk 1 above)
Comment (alxtkr77) [reply storey/redis_driver.py:104]: It is intentionally - this way you keep all related keys on the same cluster node: https://redis.com/blog/redis-clustering-best-practices-with-keys/

File: storey/redis_driver.py (hunk 2 above)
Comment (alxtkr77) [reply storey/redis_driver.py:78]: ok

File: storey/redis_driver.py (hunk 3 above)
Comment (alxtkr77) [reply storey/redis_driver.py]: ok

File: integration/test_redis_specific.py (hunk 4 above)
Comment (alxtkr77) [reply integration/test_redis_specific.py]: ok

File: integration/test_redis_specific.py (hunk 5 above)
Comment (alxtkr77) [reply integration/test_redis_specific.py]: ok

File: integration/test_redis_specific.py (hunk 6 above)
Comment (alxtkr77) [reply integration/test_redis_specific.py]: ok

File: integration/test_redis_specific.py (hunk 7 above)
Comment (alxtkr77) [reply integration/test_redis_specific.py]: ok

File: integration/test_redis_specific.py (hunk 8 above)
Comment (alxtkr77) [reply integration/test_redis_specific.py]: ok

File: integration/integration_test_utils.py [hunk 9]
```diff
@@ -225,7 +226,9 @@ def create_temp_redis_kv(setup_teardown_test):
     redis_client = get_redis_client(redis_fake_server=redis_fake_server)
 
     for i in range(1, 10):
-        redis_client.hmset(f'storey-test:{table_path}{i}:static', mapping={'age': f'{10 - i}', 'color': f'blue{i}'})
+        key = RedisDriver.make_key('storey-test:', table_path, i)
+        static_key= RedisDriver._static_data_key(key)
```
Comment (gtopper) [inline integration/integration_test_utils.py]:
```suggestion
        static_key = RedisDriver._static_data_key(key)
```

File: integration/test_redis_specific.py [hunk 10]
```diff
@@ -25,46 +27,69 @@ def redis():
...
+        controller.terminate()
+        termination_result = controller.await_termination()
+
+        expected_result = [
+            {'col1': 1, 'col2': '2', 'name': 1234},
+            {'col1': 1, 'col2': '2', 'name': 1234}]
+
+        assert termination_result == expected_result
+    except:
+        pass
```
Comment (gtopper) [inline integration/test_redis_specific.py]: Why are we ignoring all errors?

File: integration/test_flow_integration.py [hunk 11]
```diff
@@ -75,7 +75,8 @@ def _get_redis_kv_all_attrs(setup_teardown_test: TestContext, key: str):
     from .integration_test_utils import get_redis_client
     from storey.redis_driver import RedisDriver
     table_name = setup_teardown_test.table_name
-    redis_key = f'storey-test:{table_name}{key}:static'
+    hash_key = RedisDriver.make_key(f'storey-test:', table_name, key)
+    redis_key= RedisDriver._static_data_key(hash_key)
```
Comment (gtopper) [inline integration/test_flow_integration.py]:
```suggestion
    redis_key = RedisDriver._static_data_key(hash_key)
```

File: integration/test_redis_specific.py [hunk 12]
```diff
@@ -25,46 +27,69 @@ def redis():
...
+            NoSqlTarget(Table(table_name, driver)),
+            Complete()
+        ]).run()
+        controller.emit({'col1': 0}, 'key').await_result()
+        controller.terminate()
+        controller.await_termination()
+        
+        table_name = f'{table_name}/'
+        hash_key = RedisDriver.make_key('storey:', table_name, 'key')
+        redis_key= RedisDriver._static_data_key(hash_key)
```
Comment (gtopper) [inline integration/test_redis_specific.py]:
```suggestion
        redis_key = RedisDriver._static_data_key(hash_key)
```

File: integration/test_redis_specific.py [hunk 13]
```diff
@@ -13,6 +13,8 @@
 # limitations under the License.
 #
 import pytest
+import random
+import string
```
Comment (gtopper) [inline integration/test_redis_specific.py]: These imports are not used anywhere.
```suggestion
```

File: integration/test_redis_specific.py [hunk 14]
```diff
@@ -25,46 +27,69 @@ def redis():
...
+        data_strings = {}
+        for key, val in data.items():
+            if isinstance(key,bytes):
+                data_strings[key.decode('utf-8')] = val.decode('utf-8')
+            else:
+                data_strings[key] = val
 
+        assert data_strings == {"col1": '0'}
+    except:
+            pass
```
Comment (gtopper) [inline integration/test_redis_specific.py]: Why are we ignoring all errors?

File: integration/test_redis_specific.py [hunk 15]
```diff
@@ -25,46 +27,69 @@ def redis():
...
+        
+        table_name = f'{table_name}/'
+        hash_key = RedisDriver.make_key('storey:', table_name, 'key')
+        redis_key= RedisDriver._static_data_key(hash_key)
 
-    assert data_strings == {"col1": '0'}
+        data = driver.redis.hgetall(redis_key)
+        data_strings = {}
+        for key, val in data.items():
+            if isinstance(key,bytes):
```
Comment (gtopper) [inline integration/test_redis_specific.py]:
```suggestion
            if isinstance(key, bytes):
```

File: integration/test_redis_specific.py [hunk 16]
```diff
@@ -25,46 +27,69 @@ def redis():
...
-            data_strings[key.decode('utf-8')] = val.decode('utf-8')
-        else:
-            data_strings[key] = val
+        driver = RedisDriver(redis)
+        controller = build_flow([
+            SyncEmitSource(),
+            NoSqlTarget(Table(table_name, driver)),
+            Complete()
+        ]).run()
+        controller.emit({'col1': 0}, 'key').await_result()
```
Comment (gtopper) [inline integration/test_redis_specific.py]: I know it was there before, but why are we doing this? `await_termination()` should be enough. If it's not, that needs to be fixed in the implementation.

File: integration/integration_test_utils.py (hunk 9 above)
Comment (alxtkr77) [reply integration/integration_test_utils.py]: ok

File: integration/test_redis_specific.py (hunk 10 above)
Comment (alxtkr77) [reply integration/test_redis_specific.py]: right...should be 'raise' :)

File: integration/test_flow_integration.py (hunk 11 above)
Comment (alxtkr77) [reply integration/test_flow_integration.py]: ok

File: integration/test_redis_specific.py (hunk 13 above)
Comment (alxtkr77) [reply integration/test_redis_specific.py]: ok

File: integration/test_redis_specific.py (hunk 14 above)
Comment (alxtkr77) [reply integration/test_redis_specific.py]: right...should be raise :)

File: integration/test_redis_specific.py [hunk 17]
```diff
@@ -25,46 +25,69 @@ def redis():
...
+        data_strings = {}
+        for key, val in data.items():
+            if isinstance(key, bytes):
+                data_strings[key.decode('utf-8')] = val.decode('utf-8')
+            else:
+                data_strings[key] = val
 
+        assert data_strings == {"col1": '0'}
+    except:
+            raise
```
Comment (gtopper) [inline integration/test_redis_specific.py]: These two lines are redundant.
```suggestion
```

File: integration/test_redis_specific.py [hunk 18]
```diff
@@ -25,46 +25,69 @@ def redis():
...
+        controller.terminate()
+        termination_result = controller.await_termination()
+
+        expected_result = [
+            {'col1': 1, 'col2': '2', 'name': 1234},
+            {'col1': 1, 'col2': '2', 'name': 1234}]
+
+        assert termination_result == expected_result
+    except:
+        raise
```
Comment (gtopper) [inline integration/test_redis_specific.py]: These two lines are redundant.
```suggestion
```

File: integration/test_redis_specific.py [hunk 19]
```diff
@@ -25,46 +25,65 @@ def redis():
...
-    for key, val in data.items():
-        if isinstance(key,bytes):
-            data_strings[key.decode('utf-8')] = val.decode('utf-8')
-        else:
-            data_strings[key] = val
+        driver = RedisDriver(redis)
+        controller = build_flow([
+            SyncEmitSource(),
+            NoSqlTarget(Table(table_name, driver)),
+            Complete()
```
Comment (gtopper) [inline integration/test_redis_specific.py]: Please remove this Complete also (it was here for the `await_result()`).
```suggestion
```

File: storey/redis_driver.py [hunk 20]
```diff
@@ -98,19 +93,15 @@ async def inner_fn(*args, **kwargs):
...
-            if self._type is RedisType.STANDALONE:
+            try:
+                self._redis = redis.cluster.RedisCluster.from_url(self._redis_url, decode_responses=True)
+            except redis.cluster.RedisClusterException:
                 self._redis = redis.Redis.from_url(self._redis_url, decode_responses=True)
-            else:
-                self._redis = redis.cluster.RedisCluster.from_url(self._redis_url, decode_response=True)
         return self._redis
 
-    async def close(self):
```
Comment (assaf758) [inline storey/redis_driver.py:113]: why did you remove the close()?

File: storey/redis_driver.py [hunk 21]
```diff
@@ -59,7 +59,6 @@ class RedisDriver(NeedsRedisAccess, Driver):
 
     def __init__(self,
                  redis_client: Optional[Union[redis.Redis, redis.cluster.RedisCluster]] = None,
-                 redis_type: RedisType = RedisType.STANDALONE,
```
Comment (assaf758) [inline storey/redis_driver.py]: i believe class RedisType can be removed altogether.

**Review by assaf758 — CHANGES_REQUESTED** [review]

File: storey/redis_driver.py (hunk 20 above)
Comment (alxtkr77) [reply storey/redis_driver.py:113]: Closing standalone redis connection in practice does nothing, but closing clustered one closes connections to each of the instances leaving redis object in "zombi" state. Trying reuse the object later causes exception

File: integration/test_redis_specific.py (hunk 19 above)
Comment (alxtkr77) [reply integration/test_redis_specific.py]: ok

File: storey/redis_driver.py (hunk 21 above)
Comment (alxtkr77) [reply storey/redis_driver.py]: ok

**Review by gtopper — APPROVED** [review]
✅

**Review by assaf758 — APPROVED** [review]
